BaikalDataGenerator/Source/render.cpp

```diff
@@ -0,0 +1,386 @@
+        {
+            if (platform.GetDevice(i).GetType() == CL_DEVICE_TYPE_GPU)
+            {
+                m_context = CLWContext::Create(platform.GetDevice(i));
+                device_found = true;
+                break;
+            }
+        }
+    }
+
+    if (!device_found)
+    {
+        THROW_EX("can't find device");
+    }
+
+    m_factory = std::make_unique<Baikal::ClwRenderFactory>(m_context, "cache");
+    m_renderer = m_factory->CreateRenderer(Baikal::ClwRenderFactory::RendererType::kUnidirectionalPathTracer);
+    m_controller = m_factory->CreateSceneController();
+
+    for (auto& output_info: kMultipleIteratedOutputs)
+    {
+        m_outputs.push_back(m_factory->CreateOutput(output_width, output_height));
+        m_renderer->SetOutput(output_info.type, m_outputs.back().get());
+    }
+    for (auto& output_info : kSingleIteratedOutputs)
+    {
+        m_outputs.push_back(m_factory->CreateOutput(output_width, output_height));
+        m_renderer->SetOutput(output_info.type, m_outputs.back().get());
+    }
+
+    if (!std::filesystem::exists(scene_file))
+    {
+        THROW_EX("There is no any scene file to load");
+    }
+
+    m_scene = Baikal::SceneIo::LoadScene(scene_file.string(),
+                                         scene_file.parent_path().string());
+}
+
+void Render::UpdateCameraSettings(const CameraInfo& cam_state)
+{
+    if (cam_state.aperture != m_camera->GetAperture())
+    {
+        m_camera->SetAperture(cam_state.aperture);
+    }
+
+    if (cam_state.focal_length != m_camera->GetFocalLength())
+    {
+        m_camera->SetFocalLength(cam_state.focal_length);
+    }
+
+    if (cam_state.focus_distance != m_camera->GetFocusDistance())
+    {
+        m_camera->SetFocusDistance(cam_state.focus_distance);
+    }
+
+    auto cur_pos = m_camera->GetPosition();
+    auto at = m_camera->GetForwardVector();
+    auto up = m_camera->GetUpVector();
+
+    if (cur_pos != cam_state.pos ||
+        at != cam_state.at ||
+        up != cam_state.up)
+    {
+        m_camera->LookAt(cam_state.pos, cam_state.at, cam_state.up);
+    }
+}
+
+void Render::SaveOutput(const OutputInfo& info,
+                        const std::string& name,
+                        bool gamma_correction_enabled,
+                        const std::filesystem::path& output_dir)
+{
+    OIIO_NAMESPACE_USING;
+
+    auto output = m_renderer->GetOutput(info.type);
+
+    assert(output);
+
+    auto buffer = static_cast<Baikal::ClwOutput*>(output)->data();
+
+    std::vector<RadeonRays::float3> output_data(buffer.GetElementCount());
+
+    output->GetData(output_data.data());
+
+    auto width = output->width();
+    auto height = output->height();
+
+    std::vector<float> image_data(info.chanels_num * width * height);
+
+    float* dst_row = image_data.data();
+
+    if (gamma_correction_enabled &&
+       (info.type == Renderer::OutputType::kColor) &&
+       (info.chanels_num == 3))
```

Comment:
info.chanels == 3 doesn;t required here ,I think

Reply:
The further implementation requred 3 chanels, so I think it will be safer to leave verification

Reply:
Can image with Renderer::OutputType::kColor type has another channels count ?